### PR TITLE
Updates for the Config system

### LINKED ISSUE #352: Merge PySPConfigBlock back into pyutilib.misc.config
In an effort to simplify / standardize configuration / options in Pyomo, we should ensure that we only use a *single* configuration management system.  The customizations / overrides in `PySPConfigBlock` and `PySPConfigValue` should be reviewed and, if appropriate, merged into `pyutilib.misc.config`.  PySP should then directly use the `pyutilib.misc.config` system.

### PR DESCRIPTION
**Pull request by jsiirola**
## Summary/Motivation:
There were a number of longstanding inconveniences in the Config system, including:
- poor support for autocompletion in IPython and IDEs
- nonintuitive behavior for `user_values()` generator (see #352)
- inconsistent internal key formatting, which made, e.g., `ConfigDict.__contains__()` appear to give inconsistent or wrong information (see #352)
- bugs in some of the Path handling
- no online documentation
- incomplete test coverage

## Changes proposed in this PR:
- numerous fixes that address the above motivation

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

**Comment by blnicho** [comment]
@jsiirola you mention issue #352 several times in the PR description but I can't find that issue. Did you mistype the number or is the issue coming from a different repo?

**Comment by jsiirola** [comment]
#352 appears to now be Pyomo/PySP#8.  Apparently when @mrmundt transferred the issue over to PySP, that removed all trace of it here.

**Comment by blnicho** [comment]
> #352 appears to now be [Pyomo/pysp#8](https://github.com/Pyomo/pysp/issues/8). Apparently when @mrmundt transferred the issue over to PySP, that removed all trace of it here.

Oops, didn't realize that when we transferred it. Should we transfer it back (and update references to #352) since it would have been closed by this PR anyway?

File: pyomo/common/tests/test_config.py [hunk 1]
```diff
@@ -1900,6 +2170,134 @@ def test_call_options(self):
...
+                 'list': ['2', 'a']}
+            )
+
+            cfg2.list.append(10)
+            self.assertEqual(
+                cfg2.value(),
+                {'int': 100, 'in': 3, 'anon': 2, 'lambda': 1,
+                 'list': ['2', 'a', '10']}
+            )
+        # No warnings due to anythong above.
```
Comment (blnicho) [inline pyomo/common/tests/test_config.py:2210]: Typo: anything

Comment (blnicho) [inline pyomo/common/tests/test_config.py:2210]: Typo: anything

**Comment by mrmundt** [comment]
> > #352 appears to now be [Pyomo/pysp#8](https://github.com/Pyomo/pysp/issues/8). Apparently when @mrmundt transferred the issue over to PySP, that removed all trace of it here.
> 
> Oops, didn't realize that when we transferred it. Should we transfer it back (and update references to #352) since it would have been closed by this PR anyway?

I am neutral on this.

File: pyomo/common/config.py [hunk 2]
```diff
@@ -1650,85 +1750,97 @@ def __setstate__(self, state):
...
 
     def __iter__(self):
-        return self._decl_order.__iter__()
+        return (self._data[key]._name for key in self._decl_order)
 
     def __getattr__(self, name):
         # Note: __getattr__ is only called after all "usual" attribute
         # lookup methods have failed.  So, if we get here, we already
         # know that key is not a __slot__ or a method, etc...
         #if name in ConfigDict._all_slots:
```
Comment (mrmundt) [inline pyomo/common/config.py:1818]: Is this commented out line still necessary?

Comment (jsiirola) [reply pyomo/common/config.py:1815]: Not strictly necessary, but I found it useful to job my memory when I was returning to this code.

**Comment by jsiirola** [comment]
@blnicho: I believe I addressed all of your comments.

**Review by blnicho — APPROVED** [review]

**Pull request merged**